core/trino-main/src/main/java/io/trino/execution/RemoteTask.java

```diff
@@ -31,6 +31,8 @@
 
     TaskStatus getTaskStatus();
 
+    int getQueueSizeAdjustment();
```

Comment:
I don't think `RemoteTask` should contain adjustment logic. I think this should be done in `NodeSelector`. 
Some proposals:
1. You could take into account average split processing time (wall time) since last batch and plan next batch size accordingly.
2. Maybe worker should actually return if task was underutilized since last `TaskUpdateRequest` update.
3. You could enhance `io.trino.server.remotetask.HttpRemoteTask#getPartitionedSplitsInfo` to also return sum of `boolean splitsQueued = unacknowledgedSplitsInfo.getWeightSum() + taskStatus.getQueuedPartitionedSplitsWeight() > 0`.

Reply:
This is work in progress—I will look into the possibility to move this into node selector.